Html logger

hvinett opened the pull request:
## Description
Adding the html logger, which would output the test results in the html format.

## Related issue
https://github.com/microsoft/vstest/issues/2065

- [x] Add Unit Tests
- [x] Add docs for the public methods
- [x] Add exception handling
- [x] Add logging

Here is the draft html
![image](https://user-images.githubusercontent.com/39301309/63015054-9df1cd00-bead-11e9-8ccf-0cbb53ea272a.png)

Commit:
First draft for the Html Logger.

Commit:
Adding the Program.cs for making the unit tests project build

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/microsoft/vstest?pullRequest=2103) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/vstest?pullRequest=2103) before we can accept your contribution.<br/><br/>:x: hvinett [sign now](https://cla.opensource.microsoft.com/microsoft/vstest?pullRequest=2103)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/microsoft/vstest?pullRequest=2103) it.</sub>

Commit:
1. Added docs for methods
2. Added remaining unit tests
3. Removed redundant spaces.

singhsarab commented:
> <html xmlns:tp="http://schemas.datacontract.org/2004/07/Microsoft.VisualStudio.TestPlatform.Extensions.HtmlLogger" xmlns:msxsl="urn:schemas-microsoft-com:xslt">

Remove this file

---
Refers to: src/Microsoft.TestPlatform.Extensions.HtmlLogger/html.html:1 in a5fe993. [](commit_id = a5fe993775723d43a72deb91d421aa3411bd3a9f, deletion_comment = False)

singhsarab commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/Html.xslt:
```diff
@@ -0,0 +1,180 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<xsl:stylesheet version="2.0"
+    
```
Remove redundant spaces.
Add section of CSS

singhsarab commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlTransformer.cs:
```diff
@@ -0,0 +1,31 @@
+        XslCompiledTransform myXslTransform;
+
+        /// <summary>
+        /// the following  function invoesthe compiled tranform and Loads the xslt file 
```
Comments need to be reviewed. Make sure they are consistent across all the files.

singhsarab commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlTransformer.cs:
```diff
@@ -0,0 +1,31 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.IO;
```
Follow the pattern as in the other files, using statements should be inside the namespace.

singhsarab commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/Microsoft.TestPlatform.Extensions.HtmlLogger.csproj:
```diff
@@ -0,0 +1,56 @@
+    <TargetFrameworks>netstandard2.0;net451</TargetFrameworks>
+    <TargetFrameworks Condition=" '$(DotNetBuildFromSource)' == 'true' ">netstandard2.0</TargetFrameworks>
+  </PropertyGroup>
+  <!--<ItemGroup>
```
Remove redundant files like this.

singhsarab commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/sample.xml:
```diff
@@ -0,0 +1,2 @@
+<TestResults xmlns="http://schemas.datacontract.org/2004/07/Microsoft.VisualStudio.TestPlatform.Extensions.HtmlLogger" xmlns:i="http://www.w3.org/2001/XMLSchema-instance"><Results><TestResult><DisplayName>TestMethod1</DisplayName><Duration>PT0.0058002S</Duration><ErrorMessage i:nil="true"/><ErrorStackTrace i:nil="true"/><FullyQualifiedName>UnitTestProject3.UnitTest1.TestMethod1</FullyQualifiedNam ...
```
Where are we using this ?

singhsarab commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/Testresult.cs:
```diff
@@ -0,0 +1,65 @@
+    public class TestResult
+    {
+        /// <summary>
+        /// Fully Qualified name of test result
```
Nit: Comments need to be formatted correctly

singhsarab commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/Testresult.cs:
```diff
@@ -0,0 +1,65 @@
+        }
+
+        /// <summary>
+        /// givest the current testresult
```
Nit: spell

singhsarab commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/TestResults.cs:
```diff
@@ -0,0 +1,53 @@
+        internal List<string> RunLevelMessageErrorAndWarning = new List<string>();
+
+        /// <summary>
+        /// List of all results in Hieracheal model
```
>Hieracheal [](start = 35, length = 10)

Nit: Spell

singhsarab commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/TestResults.cs:
```diff
@@ -0,0 +1,53 @@
+{
+    [DataContract]
+    [KnownType(typeof(TestResult))]
+    public sealed class TestResults
```
Something like TestRunDetails.
But let's discuss, this seems like a redundant class.

singhsarab commented:
> // Copyright (c) Microsoft Corporation. All rights reserved.

Rename the file name of this file, make sure it is formatted correctly.
TestResult.cs

---
Refers to: src/Microsoft.TestPlatform.Extensions.HtmlLogger/Testresult.cs:1 in a5fe993. [](commit_id = a5fe993775723d43a72deb91d421aa3411bd3a9f, deletion_comment = False)

singhsarab commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/TestRunSummary.cs:
```diff
@@ -0,0 +1,27 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+namespace Microsoft.VisualStudio.TestPlatform.Extensions.HtmlLogger
```
Nit: Add space here

singhsarab commented on test/Microsoft.TestPlatform.Extensions.HtmlLogger.UnitTests/Microsoft.TestPlatform.Extensions.HtmlLogger.UnitTests.csproj:
```diff
@@ -0,0 +1,26 @@
+  <PropertyGroup>
+    <OutputType Condition=" '$(TargetFramework)' == 'netcoreapp2.1' ">Exe</OutputType>
+    <TargetFrameworks>netcoreapp2.1;net451</TargetFrameworks>
+    <!--<AssemblyName>Microsoft.TestPlatform.Extensions.TrxLogger.UnitTests</AssemblyName>-->
```
Nit: Remove.

singhsarab commented on test/Microsoft.TestPlatform.Extensions.HtmlLogger.UnitTests/HtmlLoggerTests.cs:
```diff
@@ -0,0 +1,457 @@
+
+
+  
+
```
Nit: please scan all the file again for spacing issues.

singhsarab commented on test/Microsoft.TestPlatform.Extensions.HtmlLogger.UnitTests/HtmlLoggerTests.cs:
```diff
@@ -0,0 +1,457 @@
+                });
+        }
+
+        /// <summary>
```
Add some useful summary for the tests, repeating what's already there in the name of the test is not advisable.

Commit:
Html logger acceptance Test added
Resources files added
updated packeges.json with html logger project
Added exception handling and eqt trace logging to files

Commit:
Merge remote-tracking branch 'upstream/master' into HtmlLogger

Commit:
incresed the file count in Tesplatform.portable and Testplatform nuspec files

Commit:
Adding xlf files

Commit:
Added installing dotnet-cli in build

Commit:
Corrected defaultParamter:Logfile on HtmlLoggerTests

Commit:
Changed the hardcoded folder name of testresults

mayankbansal018 commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs:
```diff
@@ -0,0 +1,352 @@
+                xmlSerializer.WriteObject(xmlStream, TestResults);
+                xmlStream.Close();
+
+                if (htmlFilePath == null)
```
>htmlFilePath [](start = 20, length = 12)

use string.IsNullOrEmpty()

mayankbansal018 commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs:
```diff
@@ -0,0 +1,352 @@
+                var isLogFileNameParameterExists = this.parametersDictionary.TryGetValue(HtmlLoggerConstants.LogFileNameKey, out string logFileNameValue);
+                if (isLogFileNameParameterExists && !string.IsNullOrWhiteSpace(logFileNameValue))
+                {
+                    this.htmlFilePath = Path.Combine(this.TestResultsDirPath, logFileNameValue);
```
>logFileNameValue [](start = 78, length = 16)

Is it only FileName or can be FullPath

hvinett replied:
it can be full Path

mayankbansal018 commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs:
```diff
@@ -0,0 +1,352 @@
+        /// Total SkippedTests in the Results
+        /// </summary>
+        public int SkippedTests { get; private set; }
+        public string xmlFilePath { get; private set; }
```
>xmlFilePath [](start = 22, length = 11)

Public properties should Start with Capital

mayankbansal018 commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs:
```diff
@@ -0,0 +1,352 @@
+                    TestResults.RunLevelMessageErrorAndWarning.Add(e.Message);
+                    break;
+                default:
+                    Debug.Fail("htmlLogger.TestMessageHandler: The test message level is unrecognized: {0}", e.Level.ToString());
```
Trace this

mayankbansal018 commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs:
```diff
@@ -0,0 +1,352 @@
+        private void AddToParentResult( Guid parentExecutionId, TestResult testResult)
+        {
+            // Needs to be tested!
+            TestResult ParentTestResult;
```
local variable, follow camelCase

mayankbansal018 commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs:
```diff
@@ -0,0 +1,352 @@
+        private void PopulateHtmlFile()
+        {
+            fileName = String.Format(CultureInfo.CurrentCulture, "{0}_{1}_{2}", Environment.GetEnvironmentVariable("UserName"), Environment.MachineName, FormatDateTimeForRunName(DateTime.Now));
+            xmlFilePath = this.GetFilePath("xml",this.fileName);
```
make "xml", "html" as const

mayankbansal018 commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs:
```diff
@@ -0,0 +1,352 @@
+            xmlFilePath = this.GetFilePath("xml",this.fileName);
+            Stream xmlStream = null;
+
+            try
```
use Using block for filestreams [https://stackoverflow.com/questions/7524903/should-i-call-close-or-dispose-for-stream-objects](https://stackoverflow.com/questions/7524903/should-i-call-close-or-dispose-for-stream-objects)

mayankbansal018 commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs:
```diff
@@ -0,0 +1,352 @@
+            }
+            catch (XsltCompileException xslte)
+            {
+                string xslteMessage = string.Format("HtmlLogger : Failed to convert xml file to html file. Exception : {0}", xslte.ToString());
```
>xslteMessage [](start = 23, length = 12)

nit: why do we need to create a new string.

mayankbansal018 commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/Resources/Resources.resx:
```diff
@@ -0,0 +1,127 @@
+    <value>html.xslt;System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089;utf-8</value>
+  </data>
+  <data name="HtmlFilePath" xml:space="preserve">
+    <value>Path to generated html test results file : {0}</value>
```
>Path to generated  [](start = 11, length = 18)

clean it to "Html test results file : {0}"

mayankbansal018 commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs:
```diff
@@ -0,0 +1,352 @@
+
+            string htmlfilePathMessage = string.Format(CultureInfo.CurrentCulture, HtmlResource.HtmlFilePath, this.htmlFilePath);
+            EqtTrace.Info(htmlfilePathMessage);
+            ConsoleOutput.Instance.Information(false, this.htmlFilePath);
```
>htmlFilePath [](start = 59, length = 12)

htmlfilePathMessage instead of this.htmlFilePath

mayankbansal018 commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs:
```diff
@@ -0,0 +1,352 @@
+                EqtTrace.Error(xslteMessage);
+            }
+
+            string htmlfilePathMessage = string.Format(CultureInfo.CurrentCulture, HtmlResource.HtmlFilePath, this.htmlFilePath);
```
If exception was thrown, we will still print this below message

mayankbansal018 commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs:
```diff
@@ -0,0 +1,352 @@
+            {
+                string ioexMessage = string.Format("HtmlLogger : Failed to create a xml file. Exception : {0}", ioEx.ToString());
+                EqtTrace.Error(ioexMessage);
+            }
```
how will user what went wrong, look into trx logger.

mayankbansal018 commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs:
```diff
@@ -0,0 +1,352 @@
+
+        private void PopulateHtmlFile()
+        {
+            fileName = String.Format(CultureInfo.CurrentCulture, "{0}_{1}_{2}", Environment.GetEnvironmentVariable("UserName"), Environment.MachineName, FormatDateTimeForRunName(DateTime.Now));
```
please make it this.filename

mayankbansal018 commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlTransformer.cs:
```diff
@@ -0,0 +1,33 @@
+    class HtmlTransformer : IHtmlTransformer
+    {
+        StringReader xsltStringReader = new StringReader(HtmlResources.Html);
+        XslCompiledTransform myXslTransform;
```
rename myXslTransform = xslTransform

mayankbansal018 commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/Utility/Constants.cs:
```diff
@@ -0,0 +1,55 @@
+        /// </summary>
+        public const string LogFileNameKey = "LogFileName"; 
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Security", "CA2104:DoNotDeclareReadOnlyMutableReferenceTypes")]
```
try to remove these suppresser message & see if it throws any warning/error.

mayankbansal018 reviewed (changes requested):
:clock1:

Commit:
Html.xslt is added to dll

hvinett commented on scripts/build/TestPlatform.Localization.targets:
```diff
@@ -29,18 +29,21 @@
   </ItemGroup>
 
   <Target Name="Localize" Condition="'$(ProductProject)' == 'true' and @(EmbeddedResource) != ''" DependsOnTargets="_CreateLocalizeXLF;_CreateLocalizeResx" />
+    <!--<Message Importance="High" Text="resources are @(EmbeddedResource)" />
```
Needs to be changed.

Commit:
Changed the file name html.xslt to Html.xslt and Added styling in xslt

Commit:
TestResults changed to TestRunDetails and summary comments corrected

Commit:
Added clickability and added test run duration and passing percentage

Commit:
redundant code in xslt is removed and acceptanece test is corrected

Commit:
added failed results

Commit:
code refactoring using resharper

karanjitsingh commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/Html.xslt:
```diff
@@ -0,0 +1,206 @@
+        }
+        .row {
+        border: 2px solid #ffffff;
+        width:100%;
```
```
border: 2px solid #ffffff;
background-color:#ffffff;
```

are overused a lot, see if you can make this code common

karanjitsingh commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/Html.xslt:
```diff
@@ -0,0 +1,206 @@
+        cursor:pointer;
+        background-color:#bcdaf7;
+        }
+        .listRow {
```
>listRow [](start = 9, length = 7)

generally css naming convention is like `list-row`

hvinett commented on src/package/nuspec/Microsoft.TestPlatform.Portable.nuspec:
```diff
@@ -353,7 +376,8 @@
     <file src="netcoreapp2.1\ru\Microsoft.TestPlatform.VsTestConsole.TranslationLayer.resources.dll"	target="tools\netcoreapp2.1\ru\Microsoft.TestPlatform.VsTestConsole.TranslationLayer.resources.dll" />
     <file src="netcoreapp2.1\ru\Microsoft.VisualStudio.TestPlatform.Client.resources.dll"	target="tools\netcoreapp2.1\ru\Microsoft.VisualStudio.TestPlatform.Client.resources.dll" />
     <file src="netcoreapp2.1\ru\Microsoft.VisualStudio.TestPlatform.Common.resources.dll"	target="tools\netcoreapp2.1\ru\Microsoft.VisualStudio.TestPlatform.Common.resources.dll" />
-    <file src="netcoreapp2.1\Extensions\ru\Microsoft.VisualStudio.TestPlatform.Extensions.Trx.TestLogger.resources.dll"	target="tools\netcoreapp2.1\Extensions\ru\Microsoft.VisualStudio.TestPlatform.Extensions.Trx.TestLogger.resources.dll" />
```
Add these

hvinett commented on src/package/nuspec/Microsoft.TestPlatform.Portable.nuspec:
```diff
@@ -252,6 +266,7 @@
     <file src="netcoreapp2.1\cs\Microsoft.VisualStudio.TestPlatform.Client.resources.dll"	target="tools\netcoreapp2.1\cs\Microsoft.VisualStudio.TestPlatform.Client.resources.dll" />
```
Add the htmllogger in netcoreapp2.1

hvinett committed:
Merge branch 'master' into HtmlLogger

Commit:
Fixed the xslt issues and added comments.\
Fixed the Portable nuspec file

Commit:
Merge from master

Commit:
Merge branch 'HtmlLogger' of https://github.com/hvinett/vstest into HtmlLogger

singhsarab commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs:
```diff
@@ -0,0 +1,399 @@
+                case TestOutcome.NotFound:
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException();
```
just have a default which breaks.

singhsarab commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs:
```diff
@@ -0,0 +1,414 @@
+                    ex.ToString());
+                ConsoleOutput.Instance.Error(false, string.Concat(HtmlResource.HtmlLoggerError), ex.Message);
+                return;
+            }
```
Add tests for these

singhsarab commented on test/Microsoft.TestPlatform.Extensions.HtmlLogger.UnitTests/HtmlLoggerTests.cs:
```diff
@@ -0,0 +1,422 @@
+            {
+                DisplayName = "def",
+                ErrorMessage = "error message",
+                ErrorStackTrace = "Error stack trace",
```
Incomplete test

singhsarab commented on test/Microsoft.TestPlatform.Extensions.HtmlLogger.UnitTests/HtmlLoggerTests.cs:
```diff
@@ -0,0 +1,422 @@
+        }
+
+        [TestMethod]
+        public void TestResultHandlerShouldSetDisplayNameIfNullProperly()
```
Add 2 diff tests for this

singhsarab commented on test/Microsoft.TestPlatform.Extensions.HtmlLogger.UnitTests/HtmlLoggerTests.cs:
```diff
@@ -0,0 +1,422 @@
+        }
+
+        [TestMethod]
+        public void TestResultHandlerShouldKeepTrackOfSummary()
```
Add individual tests

Commit:
unit tests refactored

Commit:
friend.cs added

Commit:
Days in GetFormattedDurationString is fixed

mayankbansal018 commented on src/Microsoft.TestPlatform.CoreUtilities/Helpers/FileHelper.cs:
```diff
@@ -132,5 +132,11 @@ public string[] GetFiles(string path, string searchPattern, SearchOption searchO
         }
+
+        /// <inheritdoc/>
+        public void Delete(string path)
```
Can you change this to TryDelete with bool return, In case file does not exists this API will throw.

mayankbansal018 commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs:
```diff
@@ -0,0 +1,410 @@
+        /// </summary>
+        /// <param name="duration"></param>
+        /// <returns></returns>
+        internal string GetFormattedDurationString(TimeSpan duration)
```
>GetFormattedDurationString [](start = 24, length = 26)

make it private

hvinett replied:
i want to test the GetFormattedDurationString function.

mayankbansal018 commented on src/Microsoft.TestPlatform.Extensions.HtmlLogger/HtmlLogger.cs:
```diff
@@ -0,0 +1,410 @@
+
+            TestResultsDirPath = testResultsDirPath;
+            TestRunDetails = new TestRunDetails();
+            Results = new ConcurrentDictionary<Guid, ObjectModel.TestResult>();
```
>Results [](start = 12, length = 7)

do we need concurrent dictionaries, I'm not sure if multiple threads are trying to add to results.

hvinett commented:
@dotnet-bot test

RehanSaeed commented:
How do we make use of the new HTML logger? The docs need updating:

https://github.com/Microsoft/vstest-docs/blob/master/docs/report.md#syntax-of-default-loggers

singhsarab commented:
@RehanSaeed Thanks. the docs have been updated.

RehanSaeed commented:
Is there a way to use it with `dotnet test` too? Executing this doesn't seem to work after upgrading my xUnit project to include 16.3.0 of the test SDK.

```bash
dotnet test --logger 'html;LogFileName=foo.html'
```

singhsarab commented:
@RehanSaeed This will be part of the dotnet sdk, and the 3.1.xx release will have it, it is still not out yet.

RehanSaeed commented:
Cool! Is the syntax to support multiple loggers like this?

```bash
dotnet test --logger trx;LogFileName=foo.trx --logger html;LogFileName=foo.html
```

dracos1993 commented:
How to use this html logger with runsettings file? I tried via cli and it works, but no luck through the config file.

`<LoggerRunSettings>
        <Loggers>
            <Logger friendlyName="html">
                <Configuration>
                    <LogFileName>test.html</LogFileName>
                </Configuration>
            </Logger>
        </Loggers>
</LoggerRunSettings>`

ViktorHofer commented:
@nohwnd can you please check if it's possible to enable the html logger via the .runsettings file? Thanks

nohwnd commented:
@RehanSaeed yes that works, but you should put those options in quotes if you run them from bash / powershell because ; terminates the command and you would end up evaluating only half of the command. Or in powershell you can use the parsing terminator `--%`: 

```shell
# cmd / bash / powershell
dotnet test --logger "trx;LogFileName=foo.trx" --logger "html;LogFileName=foo.html"

# or another way in powershell
dotnet test --% --logger trx;LogFileName=foo.trx --logger html;LogFileName=foo.html
```



@dracos1993 

This is what you'd do from settings, I actually just copied it from the log and removed unneeded lines: 

```shell
dotnet test --% --logger trx;LogFileName=foo.trx --logger html;LogFileName=foo.html --diag c:\temp\log.txt
```

```xml
<RunSettings>
  <RunConfiguration>
  </RunConfiguration>
  <LoggerRunSettings>
    <Loggers>
      <Logger friendlyName="trx" enabled="True">
        <Configuration>
          <LogFileName>foo.trx</LogFileName>
        </Configuration>
      </Logger>
      <Logger friendlyName="html" enabled="True">
        <Configuration>
          <LogFileName>foo.html</LogFileName>
        </Configuration>
      </Logger>
    </Loggers>
  </LoggerRunSettings>
</RunSettings>

```
